File: 0890-find-and-replace-pattern/0890-find-and-replace-pattern.cpp
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string>ans;
        pattern = normalize(pattern);
        for(string w:words) {
            if(pattern == normalize(w)) {
                ans.push_back(w);
            }
        }
        return ans;
    }
    
    string normalize(string &str) {
        unordered_map<char, char>mp;
        int cnt = 0;
        string ans = "";
        for(auto c:str) {
            if(mp.find(c) != mp.end()) {
                ans += mp[c];
            }
            else {
                char ch = ('a' + cnt);
                ans += ch;
                cnt++;
                mp[c] = ch;
            }
        }
        return ans;
    }
};

File: 0890-find-and-replace-pattern/0890-find-and-replace-pattern.java
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> ans = new ArrayList<>();
        String pat = normalize(pattern);
        for(String str:words) {
            if(pat.equals(normalize(str))) {
                ans.add(str);
            }
        }
        return ans;
    }
    String normalize(String str) {
        HashMap<Character, Character>mp = new HashMap<>();
        int cnt = 0;
        StringBuilder ans = new StringBuilder();
        for(char c:str.toCharArray()) {
            if(mp.containsKey(c)) {
                ans.append(mp.get(c));
            }
            else {
                char ch = (char)('a' + cnt);
                ans.append(ch);
                mp.put(c, ch);
                cnt++;
            }
        }
        return ans.toString();
    }
}